src/Skia/Avalonia.Skia/FormattedTextImpl.cs

```diff
@@ -108,135 +164,400 @@ public Size Measure()
 
         public void SetForegroundBrush(IBrush brush, int startIndex, int length)
         {
-            // TODO: we need an implementation here to properly support FormattedText
+            var key = new FBrushRange(startIndex, length);
+            if (brush == null)
+            {
+                if (_foregroundBrushes.ContainsKey(key))
+                    _foregroundBrushes.Remove(key);
+            }
+            else
+            {
+                _foregroundBrushes[key] = brush;
+            }
         }
 
-        void Rebuild()
+        public override string ToString()
         {
-            var length = _text.Length;
+            return _text;
+        }
 
-            _lines.Clear();
+        internal void Draw(SKCanvas canvas, SKPoint origin,
+                            DrawingContextImpl.PaintWrapper foreground,
+                            Func<IBrush, Size, DrawingContextImpl.PaintWrapper> brushFactory)
```

Comment:
Again, we'd better avoid passing delegates here and there in rendering code. It gets called a *lot* of times